src/components/timeline/TrackScreenshots.js

```diff
@@ -0,0 +1,261 @@
+          screenshotIndex = i;
+        } else {
+          break;
+        }
+      }
+      const { url, windowWidth, windowHeight } = screenshots.data[
+        screenshotIndex
+      ];
+      const scaledImageWidth = TRACK_HEIGHT * windowWidth / windowHeight;
+      images.push(
+        <div
+          className="timelineTrackScreenshotImgContainer"
+          style={{ left, width: imageContainerWidth }}
+          key={left}
+        >
+          {/* The following image is centered and cropped by the outer container. */}
+          <img
+            className="timelineTrackScreenshotImg"
+            src={thread.stringTable.getString(url)}
+            style={{
+              width: scaledImageWidth,
+              height: TRACK_HEIGHT,
+            }}
+          />
+        </div>
+      );
+    }
+
+    return images;
+  }
+
+  renderHoverPreview() {
+    const { pageX, offsetX, containerTop } = this.state;
+    const {
+      screenshots,
+      thread,
+      isMakingPreviewSelection,
+      width,
+      overlayElement,
+    } = this.props;
+    if (isMakingPreviewSelection || offsetX === null || pageX === null) {
+      return null;
+    }
+    const screenshotIndex = this.findScreenshotAtMouse(offsetX);
+    if (screenshotIndex === null) {
+      return null;
+    }
+    const { url, windowWidth, windowHeight } = screenshots.data[
+      screenshotIndex
+    ];
+
+    // Compute the hover image's thumbnail size.
+    let hoverHeight = HOVER_HEIGHT;
+    let hoverWidth = HOVER_HEIGHT / windowHeight * windowWidth;
+
+    if (hoverWidth > HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO) {
+      // This is a really wide image, limit the height so it lays out reasonably.
+      hoverWidth = HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO;
+      hoverHeight = hoverWidth / windowWidth * windowHeight;
+    }
+    // Set the top so it centers around the track.
+    const top = containerTop + (TRACK_HEIGHT - hoverHeight) * 0.5;
+    const left =
+      offsetX + hoverWidth * 0.5 > width
+        ? // Stick the hover image on to the right side of the container.
+          pageX - offsetX + width - hoverWidth * 0.5
+        : // Center the hover image around the mouse.
+          pageX;
+
+    return createPortal(
+      <div className="timelineTrackScreenshotHover" style={{ left, top }}>
+        <img
+          className="timelineTrackScreenshotHoverImg"
+          src={thread.stringTable.getString(url)}
+          style={{
+            height: hoverHeight,
+            width: hoverWidth,
+          }}
+        />
+      </div>,
+      overlayElement
+    );
+  }
+
+  _handleMouseOut = () => {
+    this.setState({
+      offsetX: null,
+      pageX: null,
+      containerTop: null,
+    });
+  };
+
+  _handleMouseMove = (event: SyntheticMouseEvent<HTMLDivElement>) => {
+    const { top, left } = event.currentTarget.getBoundingClientRect();
+    this.setState({
+      pageX: event.pageX,
+      offsetX: event.pageX - left,
```

Comment:
I have FUD around accessing native event properties, so I'll probably keep it this way since it's fairly simple, and I don't think the performance will be too bad.